### PR TITLE
feat(botocore): expand support for bedrock

### PR DESCRIPTION
**Pull request by Yun-Kim**
This PR adds tracing support for the `bedrock-runtime` service in botocore. Specifically, this adds tracing support for the `invoke_model` and `invoke_model_with_response_stream` operations in bedrock, and attaches prompt/response tags onto the associated bedrock spans.

### Notes for Reviewers
**The majority of the lines of code changed in this PR are from snapshot test files, riot requirement lockfiles, and test cassette data files. The real number of lines of code changed is closer to ~300 (excluding tests).**

The major files that require reviewing are:
- `ddtrace/contrib/botocore/__init__.py`
- `ddtrace/contrib/botocore/patch.py`
- `ddtrace/contrib/botocore/services/bedrock.py`
- `releasenotes/notes/feat-bedrock-8cf3f9460a529ac6.yaml`
- and `tests/contrib/botocore/test_bedrock.py`.

### Span tagging
The following request tags will be set on Bedrock spans:
Screenshot 2024-01-09 at 3 11 40 PM

The following response tags will be set on Bedrock spans:
Screenshot 2024-01-09 at 3 12 07 PM

### Limitations

Note that AI21 does not have any models that support streamed response streaming, so any requests using that model provider will not have prompt/response tagging.
Bedrock also offers a `stability` provider list of models which are image-modality based. This PR does not include support for those models, but can be added in the future.
Note: the responses returned are in streamed responses, which means that if the user does not fully consume the response, then the span will not be finished. There is the possibility of a memory leak if the number of unfinished spans gets very large, but this is following the wsgi integration example and acceptable for now.

### Future Steps

This PR only introduces tracing support for Bedrock operations. Future PRs will be completed to send prompt/response payloads to LLMObs.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (sabrenner) [inline ddtrace/contrib/botocore/services/bedrock.py:173]: Should this also be turned into a list with the condition that it already isn't one, as is done in `_extract_streamed_response`? Looks like in the case of `cohere` it might already be a list.

**Review by sabrenner — APPROVED** [review]
The contents of `ddtrace/contrib/botocore/services/bedrock.py` LGTM! Just left a small question, but not blocking on it.

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:173]: Good catch! Will update.

File: ddtrace/contrib/botocore/__init__.py
Comment (Yun-Kim) [inline ddtrace/contrib/botocore/__init__.py:30]: TODO: add mlobs as codeowner for bedrock/services.py

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [inline ddtrace/contrib/botocore/services/bedrock.py:23]: Note: need to look at https://github.com/DataDog/dd-trace-py/blob/main/ddtrace/tracing/trace_handlers.py#L78 as an example

Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:23]: Tested locally, this is a non-issue as checking for `len()` or `streaming_body.__len__` results in an attribute error without having to override `__getattribute__`.

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:27]: nit: It would be nice if TracedBotocoreStreamingBody methods had doc strings and typing

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:54]: We should move the span finish to a "finally" block

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:69]: Move this to a finally block. This way if the iterable does not finish and is garbage collected then it will close the span

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:81]: nit: we should use f strings in many places in the method. They are more efficient and easier to read

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:155]: Are completions and results guaranteed to have at least one element? Should we check the size of the element first?

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:175]: Could the text be a tuple?

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:198]: Should we check the size of streamed_body first?

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (mabdinur) [inline ddtrace/contrib/botocore/services/bedrock.py:297]:
```suggestion
        raise
    finally:
        bedrock_span.finish()
```

File: riotfile.py
Comment (mabdinur) [inline riotfile.py:1352]: We should leave a comment to explain why we need a min version of 3.8 here

**Review by mabdinur — APPROVED** [review]
left some nits but overall LGTM

File: ddtrace/contrib/botocore/patch.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/patch.py:164]: Good catch! Yeah, we'll still enable tracing if llmobs is disabled.

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:54]: This won't work, since `read(n_bytes)` can read only up to `n_bytes` and not fully consume the stream. We don't want to finish the span always when this method is called, only when the stream is fully read or on error.

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:155]: yup, the results are guaranteed to be this format. What I can do instead is to add error handling in case of index/attribute error, and return None on error.

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:175]: Nope, it'll always be a str or list

File: ddtrace/contrib/botocore/services/bedrock.py
Comment (Yun-Kim) [reply ddtrace/contrib/botocore/services/bedrock.py:297]: We can't finish the span on all cases here, since the user has to consume the body before we can finish it.

**Pull request merged**
